### PR TITLE
razorpay-js

### PR DESCRIPTION
**Pull request by cb-pravinrajmohan**

### CODE REVIEW

File: php/razorpay_js/app.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ */
+ const CB_AUTH_TOKEN = Buffer.from(process.env.CB_API_KEY + ":").toString('base64');
+ const CB_AUTH_HEADER = `Basic ${CB_AUTH_TOKEN}`;
+
+ const RP_AUTH_TOKEN = Buffer.from(process.env.RAZORPAY_KEY + ":" + process.env.RAZORPAY_KEY_SECRET).toString('base64');
+ const RP_AUTH_HEADER = `Basic ${RP_AUTH_TOKEN}`;
+
+/**
+ * [ @Chargebee ]
+ * Used to create chargebee's estimation on subscription for items
```
Comment (cb-nilotpalr) [inline php/razorpay_js/app.js:31]: Used to create an estimate for a subscription(Product Catalog 2.0)
Refer to this [API](https://apidocs.chargebee.com/docs/api/estimates?prod_cat_ver=2&lang=curl#estimate_for_creating_a_subscription).

File: php/razorpay_js/app.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            res.status(apiResult.status).json(apiResult.data);
+        })
+        .catch(function (error) {
+            console.log(error.response.data);
+        });
+})
+
+/**
+ * [ @Razorpay ]
+ * Used to create customer on razorpay
```
Comment (cb-nilotpalr) [inline php/razorpay_js/app.js:56]: Used to create a customer in Razorpay.
Refer to this [API](http://API_LINK).

@cb-pravinrajmohan Kindly add the link for the respective API.

File: php/razorpay_js/app.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            res.status(apiResult.status).json(apiResult.data);
+        })
+        .catch(function (error) {
+            console.log(error.response.data);
+        });
+})
+
+/**
+ * [ @Razorpay ]
+ * Used to create order on razorpay
```
Comment (cb-nilotpalr) [inline php/razorpay_js/app.js:81]: Used to create an order in Razorpay.
Refer to this [API](http://API_LINK).

@cb-pravinrajmohan Kindly add the link for the respective API.

File: php/razorpay_js/app.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            res.status(apiResult.status).json(apiResult.data);
+        })
+        .catch(function (error) {
+            console.log(error.response.data);
+        });
+})
+
+/**
+ * [ @Chargebee ]
+ * Used to create subscription in chargebee site
```
Comment (cb-nilotpalr) [inline php/razorpay_js/app.js:106]: Used to create a subscription(Product Catalog 2.0) for an existing customer within Chargebee
Refer to this [API](http://api_link/).

@cb-pravinrajmohan Kindly add the link for the respective API.

**Review by cb-nilotpalr — CHANGES_REQUESTED** [review]
Suggested comment changes

File: php/razorpay_js/app.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+
+/**
+ * [ @Chargebee ]
+ * Used to create an estimate for a subscription(Product Catalog 2.0)
+ * Refer to this API [https://apidocs.chargebee.com/docs/api/estimates?prod_cat_ver=2&lang=curl#estimate_for_creating_a_subscription].
+ */
+app.post('/api/create_estimate', (req, res) => {
+    let data = qs.stringify(req.body);
+    let customerId = req.body.cb_customer_id
```
Comment (cb-nilotpalr) [inline php/razorpay_js/app.js:66]: Isn't it possible to use `data.cb_customer_id` here?

File: php/razorpay_js/app.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        .then(function (apiResult) {
+            res.status(apiResult.status).json(apiResult.data);
+        })
+        .catch(function (error) {
+            console.log(error.response.data);
+        });
+})
+
+/**
+ * dummy base route
```
Comment (cb-nilotpalr) [inline php/razorpay_js/app.js:166]: Is this required here?

File: php/razorpay_js/index.html [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+			color: #003eff;
+			border: 2px solid #003eff;
+			border-radius: 5px;
+			font-weight: 600;
+		}
+	</style>
+</head>
+
+<body>
+	<h1>Chargebee Razorpay Integeration Sample</h1>
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:52]: Correct the spelling for `Integeration` -> `Integration`

File: php/razorpay_js/index.html [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+		let CB_CUSTOMER_ID;
+		let paymentType;
+		async function initatePayment() {
+			try {
+				paymentType = document.getElementById("payment_type").value;
+				payButton.classList.add('disable')
+				if(!chargebeeCustomerDetail) {
+					console.info(`=============== Creating Customer ================`);
+					/**
+					 * @note You can skip this create customer part is the customer aleady exists in Chargebee Application
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:76]: You can skip this create customer part is the customer aleady exists in Chargebee Application -> `This step can be skipped if the customer already exists within Chargebee`

File: php/razorpay_js/index.html [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+		</select>
+		<button class="btn-pay" id="rzp-button1" onclick="initatePayment()">
+			Pay with Razorpay
+		</button>
+	</div>
+	<script src="https://checkout.razorpay.com/v1/checkout.js"></script>
+	<script>
+		let ITEM_PRICE_ID = "cbdemo_advanced-INR-Yearly"; // replace with your item price id
+		const payButton = document.getElementById("rzp-button1");
+		let chargebeeCustomerDetail;
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:66]: Rename variable to `chargebeeCustomerDetails`

File: php/razorpay_js/index.html [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+						"email": "testuser17@test.com"
+					}
+					chargebeeCustomerDetail = await createChargebeeCustomer(newCbCustomer);
+					if(chargebeeCustomerDetail?.customer) {
+						CB_CUSTOMER_ID = chargebeeCustomerDetail.customer.id
+					}
+				}else{
+					/** @condition for existing Customer
+					 /* chargebeeCustomerDetail = { Your existing customer detail };
+					 /* CB_CUSTOMER_ID = chargebeeCustomerDetail.customer.id
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:91]: Uncomment this line for this `else` block to make sense

File: php/razorpay_js/index.html [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+			urlencoded.append("last_name", customerDetail.last_name || '');
+			urlencoded.append("email", customerDetail.email || '');
+
+			let requestOptions = {
+				method: 'POST',
+				headers: myHeaders,
+				body: urlencoded,
+			};
+
+			let newCbCustomer = await fetch("http://localhost:3000/api/create_cb_customer", requestOptions)
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:166]: Extract the server URL out to a global variable, then build the URLs while making the request

File: php/razorpay_js/index.html [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+				urlencoded.append("cb_customer_id", CB_CUSTOMER_ID);
+
+				let requestOptions = {
+					method: "POST",
+					headers: myHeaders,
+					body: urlencoded,
+				};
+
+				let estimate = await fetch(
+					"http://localhost:3000/api/create_estimate",
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:197]: Extract the server URL out to a global variable, then build the URLs while making the request

File: php/razorpay_js/index.html [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+				payButton.classList.remove('disable')
+				console.log("Error:", err);
+			}
+		}
+
+		async function createChargebeeCustomer(customerDetail) {
+			console.info(
+					"====== Creating CB Customer =========="
+			);
+			let myHeaders = new Headers();
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:152]: Rename `myHeaders` to `headers`

File: php/razorpay_js/index.html [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+					.catch(error => console.log('error', error));
+			return newCbCustomer;
+		}
+
+		async function createCbEstimate() {
+			console.info(
+				"====== Creating CB Subscription for Items Estimation =========="
+			);
+			try {
+				let myHeaders = new Headers();
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:178]: Rename `myHeaders` to `headers`

File: php/razorpay_js/index.html [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+				console.log("error ========", error);
+			}
+		}
+
+		async function createRazorpayCustomer(customerData) {
+			console.info(
+				`======= Creating New Customer in Razorpay ========`
+			);
+			try {
+				let myHeaders = new Headers();
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:217]: Rename `myHeaders` to `headers`

File: php/razorpay_js/index.html [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+				console.log("error ========", error);
+			}
+		}
+
+		async function createOrder(customerId, estimatedAmount, currencyCode) {
+			console.info(
+				`====== Creating Order in Razorpay =======`
+			);
+			try {
+				let myHeaders = new Headers();
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:253]: Rename `myHeaders` to `headers`

File: php/razorpay_js/index.html [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+					urlencoded.token = {
+						"max_amount": 300000,  // declare your max amount
+						"expire_at": 2709971120,
+						"frequency" : "monthly"
+					}
+				}
+
+				let requestOptions = {
+					method: "POST",
+					headers: myHeaders,
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:276]: Rename `myHeaders` to `headers`

File: php/razorpay_js/index.html [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+			}
+		}
+
+		async function createOrder(customerId, estimatedAmount, currencyCode) {
+			console.info(
+				`====== Creating Order in Razorpay =======`
+			);
+			try {
+				let myHeaders = new Headers();
+				// myHeaders.append("Authorization", RP_AUTH_HEADER);
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:254]: Remove this comment if it's not required

File: php/razorpay_js/index.html [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+				};
+
+				let requestOptions = {
+					method: "POST",
+					headers: myHeaders,
+					body: JSON.stringify(urlencoded),
+				};
+
+				let newRazorpayCustomer = await fetch(
+					"http://localhost:3000/api/create_razorpay_customer",
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:233]: Extract the server URL out to a global variable, then build the URLs while making the request

File: php/razorpay_js/index.html [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+				let requestOptions = {
+					method: "POST",
+					headers: myHeaders,
+					body: JSON.stringify(urlencoded)
+				};
+
+				// console.log("url encoded", urlencoded, requestOptions)
+
+				let orderDetails = await fetch(
+					"http://localhost:3000/api/create_razorpay_order",
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:283]: Extract the server URL out to a global variable, then build the URLs while making the request

File: php/razorpay_js/index.html [hunk 21]
```diff
@@ -0,0 +1,402 @@
...
+				urlencoded.append("cb_customer_id", CB_CUSTOMER_ID);
+
+				let requestOptions = {
+					method: "POST",
+					headers: myHeaders,
+					body: urlencoded,
+				};
+
+				let subscriptionDetails = await fetch(
+					"http://localhost:3000/api/create_subscription",
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:388]: Extract the server URL out to a global variable, then build the URLs while making the request

File: php/razorpay_js/index.html [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+				alert(response.error.metadata.payment_id);
+			});
+		}
+
+		async function createSubscrition(paymentId) {
+			console.info(
+				`====== Initiating Chargebee Subscription ======`
+			);
+			try {
+				let myHeaders = new Headers();
```
Comment (cb-nilotpalr) [inline php/razorpay_js/index.html:363]: Rename `myHeaders` to `headers`

File: php/razorpay_js/package.json [hunk 23]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "razorpay-tutorial",
+  "version": "1.0.0",
+  "description": "",
```
Comment (cb-nilotpalr) [inline php/razorpay_js/package.json:4]: Add a description for this project. For ex:
> A simple node application used for making backend API calls to Chargebee & Razorpay APIs.

File: php/razorpay_js/package.json [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+  "main": "index.js",
+  "scripts": {
```
Comment (cb-nilotpalr) [inline php/razorpay_js/package.json:6]: Remove the `test` script and add a `run` script

**Review by cb-nilotpalr — CHANGES_REQUESTED** [review]

File: php/razorpay_js/app.js (hunk 6 above)
Comment (cb-pravinrajmohan) [reply php/razorpay_js/app.js:166]: No, removed it

File: php/razorpay_js/index.html (hunk 7 above)
Comment (cb-pravinrajmohan) [reply php/razorpay_js/index.html:52]: Done

File: php/razorpay_js/index.html (hunk 8 above)
Comment (cb-pravinrajmohan) [reply php/razorpay_js/index.html:76]: Done

File: php/razorpay_js/app.js (hunk 5 above)
Comment (cb-nilotpalr) [reply php/razorpay_js/app.js:66]: Okay! Thanks for the explanation.

**Review by cb-nilotpalr — APPROVED** [review]
All comments has been resolved

**Pull request merged**